File: api/src/modules/notifications/notification.controller.ts
import { Body, Controller, Get, Param, Patch, Query } from "@nestjs/common";

import { CurrentUser } from "../auth/decorators/current-user.decorator";
import { AuthenticatedUser } from "../auth/interfaces/authenticated-user.interface";
import { NotificationRecord, NotificationService } from "./notification.service";

@Controller("notifications")
export class NotificationController {
    constructor(private readonly notificationService: NotificationService) { }

    @Get()
    list(
        @CurrentUser() user: AuthenticatedUser,
        @Query("unreadOnly") unreadOnly?: string,
        @Query("limit") limit?: string,
        @Query("offset") offset?: string
    ): Promise<{
        items: NotificationRecord[];
        total: number;
        limit: number;
        offset: number;
        unreadCount: number;
    }> {
        return this.notificationService.list({
            userId: user.userId,
            unreadOnly: unreadOnly === "true",
            limit: limit ? parseInt(limit, 10) : undefined,
            offset: offset ? parseInt(offset, 10) : undefined
        });
    }

    @Get("unread-count")
    getUnreadCount(
        @CurrentUser() user: AuthenticatedUser
    ): Promise<{ unreadCount: number }> {
        return this.notificationService.getUnreadCount(user.userId);
    }

    @Patch(":id/read")
    markRead(
        @Param("id") notificationId: string,
        @CurrentUser() user: AuthenticatedUser
    ): Promise<NotificationRecord> {
        return this.notificationService.markRead(notificationId, user.userId);
    }

    @Patch("read-all")
    markAllRead(
        @CurrentUser() user: AuthenticatedUser
    ): Promise<{ updated: number }> {
        return this.notificationService.markAllRead(user.userId);
    }
}

File: api-java/src/main/java/com/illamhelp/api/notifications/NotificationController.java
package com.illamhelp.api.notifications;

import com.illamhelp.api.common.CurrentUser;
import java.util.Map;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class NotificationController {
  private final NotificationService service;

  public NotificationController(NotificationService service) {
    this.service = service;
  }

  @GetMapping("/notifications")
  public Map<String, Object> list(
      @AuthenticationPrincipal Jwt jwt,
      @RequestParam(required = false, defaultValue = "false") boolean unreadOnly,
      @RequestParam(required = false) Integer limit,
      @RequestParam(required = false) String cursor) {
    return service.list(CurrentUser.fromJwt(jwt).userId(), unreadOnly, limit, cursor);
  }

  @GetMapping("/notifications/unread-count")
  public Map<String, Integer> unreadCount(@AuthenticationPrincipal Jwt jwt) {
    return service.unreadCount(CurrentUser.fromJwt(jwt).userId());
  }

  @PatchMapping("/notifications/{id}/read")
  public Map<String, Object> markRead(@AuthenticationPrincipal Jwt jwt, @PathVariable String id) {
    return service.markRead(CurrentUser.fromJwt(jwt).userId(), id);
  }

  @PatchMapping("/notifications/read-all")
  public Map<String, Integer> markAllRead(@AuthenticationPrincipal Jwt jwt) {
    return service.markAllRead(CurrentUser.fromJwt(jwt).userId());
  }
}
